packages/aws-cdk/lib/api/accelerated-deployments.ts

```diff
@@ -0,0 +1,242 @@
+import * as cfn_diff from '@aws-cdk/cloudformation-diff';
+import * as cxapi from '@aws-cdk/cx-api';
+import { ISDK, Mode, SdkProvider } from './aws-auth';
+import { DeployStackResult } from './deploy-stack';
+import { CloudFormationStack } from './util/cloudformation';
+import { evaluateCfn } from './util/cloudformation/evaluate-cfn';
+
+/**
+ * Perform an accelerated deployment,
+ * short-circuiting CloudFormation if possible.
+ * If it's not possible to short-circuit the deployment
+ * (because the CDK Stack contains changes that cannot be deployed without CloudFormation),
+ * returns `undefined`.
+ */
+export async function performAcceleratedDeployment(
+  sdkProvider: SdkProvider, assetParams: { [key: string]: string },
+  cloudFormationStack: CloudFormationStack, stackArtifact: cxapi.CloudFormationStackArtifact,
+): Promise<DeployStackResult | undefined> {
+  const currentTemplate = await cloudFormationStack.template();
+  const stackChanges = cfn_diff.diffTemplate(currentTemplate, stackArtifact.template);
+
+  // resolve the environment, so we can substitute things like AWS::Region in the CFN expressions
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  const acceleratableChanges = findAllAcceleratableChanges(stackChanges, {
+    ...assetParams,
+    'AWS::Region': resolvedEnv.region,
+    'AWS::AccountId': resolvedEnv.account,
+  });
+  if (!acceleratableChanges) {
+    // this means there were changes to the template that cannot be short-circuited
+    return undefined;
+  }
+
+  // create a new SDK using the CLI credentials, because the default one will not work for new-style synthesis -
+  // it assumes the bootstrap deploy Role, which doesn't have permissions to update Lambda functions
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForWriting);
+  // apply the short-circuitable changes
+  await applyAllAcceleratedChanges(sdk, stackArtifact, acceleratableChanges);
+
+  return { noOp: acceleratableChanges.length === 0, stackArn: cloudFormationStack.stackId, outputs: cloudFormationStack.outputs, stackArtifact };
+}
+
+function findAllAcceleratableChanges(
+  stackChanges: cfn_diff.TemplateDiff, assetParamsWithEnv: { [key: string]: string },
+): LambdaFunctionResource[] | undefined {
+  const acceleratableResources = new Array<LambdaFunctionResource>();
+  let foundNonAcceleratableChange = false;
+  stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const lambdaFunctionShortCircuitChange = isAcceleratedLambdaFunctionChange(logicalId, change, assetParamsWithEnv);
+    if (lambdaFunctionShortCircuitChange === false) {
+      foundNonAcceleratableChange = true;
```

Comment:
We should likely break at this point since if there is one non-acceleratable change, we are done anyway - right?

Reply:
Perhaps, but notice this is not a `for` loop, but a `forEachDifference()` method of `DifferenceCollection`, so we can't break from it (that's the API `@aws-cdk/cloudformation-dff` gives you).